support/langkit_support-generic_api.ads

```diff
@@ -0,0 +1,76 @@
+------------------------------------------------------------------------------
+--                                                                          --
+--                                 Langkit                                  --
+--                                                                          --
+--                     Copyright (C) 2014-2021, AdaCore                     --
+--                                                                          --
+-- Langkit is free software; you can redistribute it and/or modify it under --
+-- terms of the  GNU General Public License  as published by the Free Soft- --
+-- ware Foundation;  either version 3,  or (at your option)  any later ver- --
+-- sion.   This software  is distributed in the hope that it will be useful --
+-- but WITHOUT ANY WARRANTY;  without even the implied warranty of MERCHAN- --
+-- TABILITY  or  FITNESS  FOR A PARTICULAR PURPOSE.                         --
+--                                                                          --
+-- As a special  exception  under  Section 7  of  GPL  version 3,  you are  --
+-- granted additional  permissions described in the  GCC  Runtime  Library  --
+-- Exception, version 3.1, as published by the Free Software Foundation.    --
+--                                                                          --
+-- You should have received a copy of the GNU General Public License and a  --
+-- copy of the GCC Runtime Library Exception along with this program;  see  --
+-- the files COPYING3 and COPYING.RUNTIME respectively.  If not, see        --
+-- <http://www.gnu.org/licenses/>.                                          --
+------------------------------------------------------------------------------
+
+--  This package and its children provide generic APIs so that programs can
+--  work with all Langkit-generated libraries.
+--
+--  Note that it is experimental at this stage, and thus not officially
+--  supported.
+
+limited private with Langkit_Support.Internal;
+
+package Langkit_Support.Generic_API is
+
+   type Language_Id is private;
+   --  Unique identifier for a Langkit-generated library
+
+   function Language_Name (Id : Language_Id) return String;
+   --  Return the name of the language that the library corresponding to ``Id``
+   --  analyzes.
+
+   type Any_Grammar_Rule is new Natural;
+   subtype Grammar_Rule is Any_Grammar_Rule range 1 .. Any_Grammar_Rule'Last;
```

Comment:
What's the point of having it be a public `Natural` rather than a private type ?

Reply:
More generally I wonder if it wouldn't be better to have those grammar rules encapsulate the language id. The fact that you can (and will) have a `Grammar_Rule` instance that is valid in two different languages but means something different is a bit unsettling to me.

Reply:
As discussed live: having a scalar allows super easy iteration & debugging of invalid values for users. It’s also possible with a private type that encapsulates the language id + the scalar, but then we need to check the language id everywhere we accept a grammar rule + the extra “rule_to_language” accessor to ease debugging on users’ side. We agreed to keep the current type and rename it `Grammar_Rule_Index`.